// @elastic/elasticsearch/esm/api/api/bulk.js
/*
 * Copyright Elasticsearch B.V. and contributors
 * SPDX-License-Identifier: Apache-2.0
 */
// This file was automatically generated by elastic/elastic-client-generator-js
// DO NOT MODIFY IT BY HAND. Instead, modify the source open api file,
// and elastic/elastic-client-generator-js to regenerate this file again.
import { errors } from '@elastic/transport';
const commonQueryParams = ['error_trace', 'filter_path', 'human', 'pretty'];
const acceptedParams = {
    bulk: {
        path: [
            'index'
        ],
        body: [
            'operations'
        ],
        query: [
            'include_source_on_error',
            'list_executed_pipelines',
            'pipeline',
            'refresh',
            'routing',
            '_source',
            '_source_excludes',
            '_source_includes',
            'timeout',
            'wait_for_active_shards',
            'require_alias',
            'require_data_stream'
        ]
    }
};
export default async function BulkApi(params, options) {
    var _a, _b;
    const operations = (_a = params.operations) !== null && _a !== void 0 ? _a : params.body;
    if (!Array.isArray(operations) || operations.length === 0) {
        throw new errors.ConfigurationError('Missing required parameter [operations]: must be a non-empty array');
    }
    const { path: acceptedPath, body: acceptedBody, query: acceptedQuery } = acceptedParams.bulk;
    const userQuery = params === null || params === void 0 ? void 0 : params.querystring;
    const querystring = userQuery != null ? { ...userQuery } : {};
    let body = (_b = params.body) !== null && _b !== void 0 ? _b : undefined;
    for (const key in params) {
        if (acceptedBody.includes(key)) {
            // @ts-expect-error
            body = params[key];
        }
        else if (acceptedPath.includes(key)) {
            continue;
        }
        else if (key !== 'body' && key !== 'querystring') {
            if (acceptedQuery.includes(key) || commonQueryParams.includes(key)) {
                // @ts-expect-error
                querystring[key] = params[key];
            }
            else {
                body = body !== null && body !== void 0 ? body : {};
                // @ts-expect-error
                body[key] = params[key];
            }
        }
    }
    let method = '';
    let path = '';
    if (params.index != null) {
        method = 'POST';
        path = `/${encodeURIComponent(params.index.toString())}/_bulk`;
    }
    else {
        method = 'POST';
        path = '/_bulk';
    }
    const meta = {
        name: 'bulk',
        pathParts: {
            index: params.index
        },
        acceptedParams: {
            path: ['index'],
            body: ['operations'],
            query: ['include_source_on_error', 'list_executed_pipelines', 'pipeline', 'refresh', 'routing', '_source', '_source_excludes', '_source_includes', 'timeout', 'wait_for_active_shards', 'require_alias', 'require_data_stream']
        }
    };
    return await this.transport.request({ path, method, querystring, bulkBody: body, meta }, options);
}
